text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+
+It could be.  But it would still need to be something that doesn't cause name
+resolution failures for other methods that people might already have written.
+
+So I don't think that the extra `()` on it would really improve things.
+
+## Why not allow writing to the discriminant?
+
+The semantics for that get really complicated, especially for `enum`s in `repr(Rust)`
+that don't have a guaranteed layout, and even more so those that get layout-optimized.
+
+Maybe one day it could be allowed, but for now this RFC sticks only things that
+can be allowed in safe code without worries.
+
+## Couldn't this be a magic macro?
+
+Sure, it could, like `offset_of!`.
+
+I don't think `enum_discriminant!(foo)` is really better than `foo.enum#discriminant`, though.
+
+It doesn't deal in tokens, and there's no special logic to apply to the scope in which
+the argument is computed.
+
+It works on a value or place, not on anything dealing tokens, nor does it affect a scope.
+
+## Why not do *\<more complex feature\>*?
+
+Privacy is the problem.
+
+If we wanted to just expose everything's discriminant to everyone, it'd be easy
+to have a trait in core that's auto-implemented for every `enum`.
+
+But to do things in a way that doesn't add a new category of major breaking change,
+that gets harder.
+
+It'd be great if we had scoped trait impls, for example, so we could do that
+in a way where it's up to the trait author how visible things get.  But that's
+a *massive* feature, so it would be nice not to block on it.
+
+Or libs-api could create a new trait and a new `derive` that's implemented using
+the same magic that today's `derive(PartialOrd)` uses.  But that's another big
+bikeshed, and doesn't even work very well for the "I'm writing my own customized
+derive" cases that just want to use the discriminant internally.
+
+The goal here is to do something easy using syntactic space that's not particularly
+valuable anyway -- if people end up almost never using this directly because there's
+a popular community `derive`, that's great.
+
+## What about `as`?
+
+While `as` *works* on field-less enums, it's not that great there either.
+
+It has the fundamental problem that you have to write out the target type that you want,
+and the wrong one will silently truncate.  This hits the same general "`as` is error-prone"
+theme that is pushing people away from using `as` to using more-specific things
+instead that are either lossless or clearer, to help avoid mistakes.
+
+If this exists, I wouldn't be surprised to see people using `foo.enum#discriminant`
+even in places where `foo as u8` works and is shorter since you don't have to think
+"what was the `repr` of this, again?" and you just get the right thing.
+
+Should the enum's declared `repr` not be the type you actually want, you can always
+use `.enum#discriminant` and *then* `as` cast it -- or hopefully `.into()` or
+something else with clearer intent -- into the type you need.
+
```

Comment:
I have a proposed alternative, something like this (some formatting etc would need to be corrected):

```suggestion
## Private method on Discriminant<T>

We could instead have the compiler generate something like this for any enum declaration:

```rust
enum Bar {}

// generated:
impl Discriminant<Bar> {
    fn discriminant(&self) -> <Bar as DiscriminantKind>::Discriminant { ... }
}
\```

This would function equivalently to a `pub(in self)` function defined by the user, callable only within the module. If we want to avoid a coherence-break of the impl, this could be implemented as synthesizing an unnameable trait that's only visible in the module:

```rust
trait BarDiscriminantAccessor { fn discriminant(&self) -> <Bar as DiscriminantKind>::Discriminant; }
impl BarDiscriminantAccessor for Discriminant<Bar> {
    fn discriminant(&self) -> <Bar as DiscriminantKind>::Discriminant { ... }
}
\```

This solution provides the same visibility/privacy as the solution proposed in the main RFC, but avoids introducing any new syntax or even particularly new concepts to the language (just auto-generated trait impls, but that's nothing a derive macro couldn't do on stable today as far as I know).
```

Reply:
Note that the impl is *not* on the enum. It's on Discriminant. So I don't think there's any compatibility concern or naming problem - you'd implement the public method on your enum.

It feels pretty unlikely that people are implementing traits on Discriminant today since you can't actually do much useful in it.